src/bsd.rs

```diff
@@ -0,0 +1,325 @@
+        unsafe { self.sin.sin_len }
+    }
+}
+
+impl From<IpAddr> for SockaddrStorage {
+    fn from(ip: IpAddr) -> Self {
+        match ip {
+            IpAddr::V4(ip) => SockaddrStorage {
+                sin: sockaddr_in {
+                #[allow(clippy::cast_possible_truncation)]
+                // `sockaddr_in` len is <= u8::MAX per `const_assert!` above.
+                sin_len: size_of::<sockaddr_in>() as u8,
+                sin_family: AF_INET_U8,
+                sin_addr: in_addr {
+                    s_addr: u32::from_ne_bytes(ip.octets()),
+                },
+                sin_port: 0,
+                sin_zero: [0; 8],
+            },
+            },
+            IpAddr::V6(ip) => SockaddrStorage {
+                sin6: sockaddr_in6 {
+                #[allow(clippy::cast_possible_truncation)]
+                // `sockaddr_in6` len is <= u8::MAX per `const_assert!` above.
+                sin6_len: size_of::<sockaddr_in6>() as u8,
+                sin6_family: AF_INET6_U8,
+                sin6_addr: in6_addr {
+                    s6_addr: ip.octets(),
+                },
+                sin6_port: 0,
+                sin6_flowinfo: 0,
+                sin6_scope_id: 0,
+            },
+            },
+        }
+    }
+}
+
+#[repr(C)]
+struct RouteMessage {
+    rtm: rt_msghdr,
+    sa: SockaddrStorage,
+}
+
+impl RouteMessage {
+    fn new(remote: IpAddr, seq: i32) -> Self {
+        let sa = SockaddrStorage::from(remote);
+        Self {
+            rtm: rt_msghdr {
+                #[allow(clippy::cast_possible_truncation)]
+                // `rt_msghdr` len + `ALIGN` is <= u8::MAX per `const_assert!` above.
+                rtm_msglen: (size_of::<rt_msghdr>() + aligned_by(sa.len().into(), ALIGN)) as u16,
+                rtm_version: RTM_VERSION_U8,
+                rtm_type: RTM_GET_U8,
+                rtm_seq: seq,
+                rtm_addrs: RTM_ADDRS,
+                ..Default::default()
+            },
+            sa,
+        }
+    }
+
+    const fn version(&self) -> u8 {
+        self.rtm.rtm_version
+    }
+
+    const fn seq(&self) -> i32 {
+        self.rtm.rtm_seq
+    }
+
+    const fn kind(&self) -> u8 {
+        self.rtm.rtm_type
+    }
+
+    const fn len(&self) -> usize {
+        self.rtm.rtm_msglen as usize
+    }
+}
+
+impl From<&RouteMessage> for &[u8] {
+    fn from(value: &RouteMessage) -> Self {
+        debug_assert!(value.len() >= size_of::<Self>());
+        unsafe { slice::from_raw_parts(ptr::from_ref(value).cast(), value.len()) }
+    }
+}
+
+impl From<Vec<u8>> for rt_msghdr {
+    fn from(value: Vec<u8>) -> Self {
+        debug_assert!(value.len() >= size_of::<Self>());
+        unsafe { ptr::read_unaligned(value.as_ptr().cast()) }
+    }
+}
+
+fn if_index(remote: IpAddr) -> Result<u16> {
+    // Open route socket.
+    let mut fd = RouteSocket::new(PF_ROUTE, AF_UNSPEC)?;
+
+    // Send route message.
+    let query = &RouteMessage::new(remote, fd.as_raw_fd());
```

Comment:
Why are you taking a reference here?  I don't see anywhere that you wouldn't be able to take the value.